doc/sfskey.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.14
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "SFSKEY 1"
.TH SFSKEY 1 "2004-10-16" "SFS 0.8pre" "SFS 0.8pre"
.SH "NAME"
sfskey \- SFS key manager
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
sfskey [\-S \fIsock\fR] [\-p \fIpwfd\fR] 
\&\fIcommand\fR [\fIarg\fR ...]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBsfskey\fR command performs a variety of key management tasks,
from generating and updating keys to controlling users' \s-1SFS\s0 agents.  The
general usage for \fBsfskey\fR is:
.PP
sfskey [\-S \fIsock\fR] [\-p \fIpwfd\fR] \fIcommand\fR [\fIarg\fR ...]
.PP
\&\fB\-S\fR specifies a \s-1UNIX\s0 domain socket \fBsfskey\fR can use to
communicate with your \fBsfsagent\fR socket.  If \fIsock\fR begins
with \fB\-\fR, the remainder is interpreted as a file descriptor number.
The default is to use the environment variable \fB\s-1SFS_AGENTSOCK\s0\fR if
that exists.  If not, \fBsfskey\fR asks the file system for a
connection to the agent.
.PP
The \fB\-p\fR option specifies a file descriptor from which
\&\fBsfskey\fR should read a passphrase, if it needs one, instead of
attempting to read it from the user's terminal.  This option may be
convenient for scripts that invoke \fBsfskey\fR.  For operations
that need multiple passphrases, you must specify the \fB\-p\fR option
multiple times, once for each passphrase.
.PP
In \s-1SFS\s0 0.7, two-party proactive Schnorr signatures (2\-Schnorr for short)
are supported in addition to Rabin signatures.  One half of the  2\-Schnorr
key is stored on the designated signature sever, while the other is stored
locally to file, or remotely via \s-1SRP\s0.  Unlike Rabin keys, 2\-Schnorr keys
can fail to load when a signature server becomes unavailable.  For this 
reason, \fBsfskey\fR supports multiple private-key shares that correspond
to the same public key; this way, a user can maintain a series of backup
signature servers in case his primary server becomes unavailable.  By
default, \fBsfskey\fR never stores both halves of a 2\-Schnorr key
to the same machine, so as to enforce key sharing.  To this effect,
2\-Schnorr employs special \fBsfskey\fR commands\*(--\fBsfskey 2gen\fR 
and \fBsfskey 2edit\fR. 
.PP
As of \s-1SFS\s0 0.7, there is a new convention for saving and naming private
keys.  By default, keys will be stored locally in \fI$HOME/.sfs/authkeys\fR,
and will be in the following forms:
.PP
.Vb 2
\&    user@host1#n
\&    user@host1#n,p.host2,m
.Ve
.PP
The first form is for standard Rabin keys.  The second is for 2\-Schnorr
proactive signature keys.  In the above examples, \fIhost1\fR is the
the full hostname of the generating host, \fIn\fR is the public key
version, \fIp\fR is the priority of the signing host (1 is the highest)
\&\fIhost2\fR is the full hostname of the signing host, and \fIm\fR is the
private key version.  
.PP
In general, these details can remain hidden, in that the symbolic link
\&\fI$HOME/.sfs/identity\fR points to the most recent key generated in
\&\fI$HOME/.sfs/authkeys\fR, and most \fBsfskey\fR commands have
reasonable defaults.  However, there is a command-line system for
accessing and generating specific keys.  A blank keyname and the
special keyname \fI#\fR refer to the default key
\&\fI$HOME/.sfs/identity\fR during key access and the next available
key during key generation.  Keynames containing a \fI#\fR character
but not containing a \fI/\fR character are assumed to refer to keys
in the \fI$HOME/.sfs/authkeys\fR directory.  When given files of the
form \fIprefix\fR\fI#\fR, \fBsfskey\fR looks in the default
directory for the most recent key with the given \fIprefix\fR during key
access, and the next available key with the given \fIprefix\fR during key
generation.  For keys of the form \fIname\fR\fI#\fR\fIsuffix\fR,
\&\fBsfskey\fR will look in the \fI$HOME/.sfs/authkeys\fR directory
for keys that match the given name exactly. \fBsfskey\fR treats
keys with \fI/\fR characters as regular files; it treats keys that
contain \fI@\fR characters but no \fI#\fR characters as keys stored
on remote machines.
.PP
Finally, one should note that \s-1SFS\s0 keys have both a \fIkeyname\fR 
and also a \fIkeylabel\fR.  \fBsfskey\fR uses the former to
retrieve keys from the local file system or from remote servers.  The latter
is less important; the \fIkeylabel\fR is stored internally in the 
private key, and is shown in the output of the \fBsfskey list\fR
command.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "sfskey add [\-t [hrs:]min] [\fIkeyname\fR]" 4
.IX Item "sfskey add [-t [hrs:]min] [keyname]"
.PD 0
.IP "sfskey add [\-t [hrs:]min] [\fIuser\fR]@\fIhostname\fR" 4
.IX Item "sfskey add [-t [hrs:]min] [user]@hostname"
.PD
The \fBadd\fR command loads and decrypts a private key, and gives
the key to your agent.  Your agent will use it to try to authenticate
you to any file systems you reference.  The \fB\-t\fR option specifies
a timeout after which the agent should forget the private key.
.Sp
In the first form of the command, the key indicated by \fIkeyname\fR
is loaded.  If \fIkeyname\fR is omitted, or \fI#\fR is supplied, then
the default key is \fI$HOME/.sfs/identity\fR. If the
key supplied is a 2\-Schnorr key, then \fBsfskey add\fR will
attempt to load backup keys should the primary key fail due to an
unavailable signature server.
.Sp
The second form of the command fetches a private key over
the network using the
\&\s-1SRP\s0 (\fIhttp://srp.stanford.edu/\fR) protocol.  \s-1SRP\s0
lets users establish a secure connection to a server without remembering
its public key.  Instead, to prove their identities to each other, the
user remembers a secret password and the server stores a one-way
function of the password (also a secret).  \s-1SRP\s0 addresses the fact that
passwords are often poorly chosen; it ensures that an attacker
impersonating one of the two parties cannot learn enough information to
mount an off-line password guessing attack\*(--in other words, the
attacker must interact with the server or user on every attempt to guess
the password.
.Sp
The \fBsfskey update\fR, \fBsfskey register\fR, 
\&\fBsfskey 2gen\fR and \fBsfskey 2edit\fR commands let users
store their private keys on servers, and retrieve them using the
\&\fBadd\fR command.  The private key is stored in encrypted form,
using the same password as the \s-1SRP\s0 protocol (a safe design as the server
never sees any password-equivalent data).
.Sp
Because the second form of \fBsfskey add\fR establishes a secure
connection to a server, it also downloads the servers HostID securely
and creates a symbolic link from \fI/sfs/\fR\fIhostname\fR to the
server's self-certifying pathname.
.Sp
When invoking \fBsfskey add\fR with the \s-1SRP\s0 syntax, \fBsfskey\fR
will ask for the user's password with a prompt of the following form:
.Sp
Passphrase for \fIuser\fR@\fIservername\fR/\fInbits\fR:
.Sp
\&\fIuser\fR is simply the username of the key being fetched from the
server.  \fIservername\fR is the name of the server on which the user
registered his \s-1SRP\s0 information.  It may not be the same as the
\&\fIhostname\fR argument to \fBsfskey\fR if the user has supplied a
hostname alias (or \s-1CNAME\s0) to \fBsfskey add\fR.  Finally, \fInbits\fR
is the size of the prime number used in the \s-1SRP\s0 protocol.  Higher values
are more secure; 1,024 bits should be adequate.  However, users should
expect always to see the same value for \fInbits\fR (otherwise, someone
may be trying to impersonate the server).
.IP "sfskey certclear" 4
.IX Item "sfskey certclear"
Clears the list of certification programs the agent runs.
See \fIcertprog\fR, for more details on certification programs.
.IP "sfskey certlist [\-q]" 4
.IX Item "sfskey certlist [-q]"
Prints the list of certification programs the agent runs.
See \fIcertprog\fR, for more details on certification programs.
.IP "sfskey certprog [\-p \fIprefix\fR] [\-f \fIfilter\fR] [\-e \fIexclude\fR] \fIprog\fR [\fIarg\fR ...]" 4
.IX Item "sfskey certprog [-p prefix] [-f filter] [-e exclude] prog [arg ...]"
The \fBcertprog\fR command registers a command to be run to lookup
\&\fIHostID\fRs on the fly in the \fI/sfs\fR directory.  This mechanism can be
used for \fBdynamic server authentication\fR\-\-running code to lookup
\&\fIHostID\fRs on\-demand.  When you reference the file
\&\fI/sfs/\fR\fIprefix\fR\fI/\fR\fIname\fR, your agent will run the command:
.Sp
\&\fIprog\fR \fIarg\fR ... \fIname\fR
.Sp
If the program succeeds and prints \fIdest\fR to its standard output,
the agent will then create a symbolic link:
.Sp
/sfs/\fIprefix\fR/\fIname\fR \-> \fIdest\fR
.Sp
The \fB\-p\fR flag can be omitted, and the link is
\&\fB/sfs/\fR\fIname\fR\fB \-\fR >\fIdest\fR.  \fIprefix\fR can be more than one
directory deep (i.e., a series of path components separated by
\&\fB/\fR).  If so, the first certification program whose prefix matches
at the beginning of \fIprefix\fR is run.  The remaining path components
are passed to \fIprog\fR.  For example:
.Sp
.Vb 1
\&  NEED EXAMPLE
.Ve
.Sp
\&\fIfilter\fR is a perl-style regular expression.  If it is specified,
then \fIname\fR must contain it for the agent to run \fIprog\fR.
\&\fIexclude\fR is another regular expression, which, if specified,
prevents the agent from running \fIprog\fR on \fIname\fRs that contain
it (regardless of \fIfilter\fR).
.Sp
The program \fBdirsearch\fR can be used with \fBcertprog\fR to
configure \fBcertification paths\fR\-\-lists of directories in which to
look for symbolic links to \fIHostID\fRs.  The usage is:
.Sp
dirsearch [\-clpq] \fIdir1\fR [\fIdir2\fR ...] \fIname\fR
.Sp
\&\fBdirsearch\fR searches through a list of directories \fIdir1\fR,
\&\fIdir2\fR, ... until it finds one containing a file called
\&\fIname\fR, then prints the pathname \fIdir\fR\fB/\fR\fIname\fR.  If it
does not find a file, \fBdirsearch\fR exits with a non-zero exit
code.  The following options affect \fBdirsearch\fR's behavior:
.RS 4
.IP "\-c" 4
.IX Item "-c"
Print the contents of the file to standard output, instead of its
pathname.
.IP "\-l" 4
.IX Item "-l"
Require that \fIdir\fR\fB/\fR\fIname\fR be a symbolic link, and print
the path of the link's destination, rather than the path of the link
itself.
.IP "\-p" 4
.IX Item "-p"
Print the path \fIdir\fR\fB/\fR\fIname\fR.  This is the default
behavior anyway, so the option \fB\-p\fR has no effect.
.IP "\-q" 4
.IX Item "-q"
Do not print anything.  Exit abnormally if \fIname\fR is not found in
any of the directories.
.RE
.RS 4
.Sp
As an example, to lookup self-certifying pathnames in the directories
\&\fI$HOME/.sfs/known_hosts\fR and \fI/mit\fR, but only accepting links
in \fI/mit\fR with names ending \fI.mit.edu\fR, you might execute the
following commands:
.Sp
.Vb 2
\&  % sfskey certprog dirsearch $HOME/.sfs/known_hosts
\&  % sfskey certprog -f '\e.mit\e.edu$' /mnt/links
.Ve
.RE
.IP "sfskey confclear" 4
.IX Item "sfskey confclear"
Clears the confirmation program that the agent runs.
See \fIconfprog\fR, for more details on confirmation programs.
.IP "sfskey conflist [\-q]" 4
.IX Item "sfskey conflist [-q]"
Prints the confirmation program that the agent runs.
See \fIconfprog\fR, for more details on confirmation programs.
.IP "sfskey confprog \fIprog\fR [\fIarg\fR ...]" 4
.IX Item "sfskey confprog prog [arg ...]"
The \fBconfprog\fR command registers a command to be run by the agent
when it receives an authentication request.  The agent provides the program
with the following command line arguments: the machine making the request,
the machine that the requestor wants to access, the service (e.g., file
system, remote execution facility), the current key that the agent will try
signing with, and a list of all of the keys that the agent has available.
If the confirmation program returns a zero exit status, the agent will
sign with the current key; otherwise, it will refuse to sign with that key
and will try the next available one.
.Sp
The confirmation program can be very simple (always answer yes, for
example), or quite complex.  \s-1SFS\s0 comes with an example confirmation program
written in Python/GTK2 (\fIconfirm.py\fR).  When called, the script can pop
up a dialog box which asks the user what he wants to do with the request.
The user has several options: reject, accept, accept and allow all futures
request from the requesting machine to access the named machine, accept and
allow access from requestor to any machine in the named machine's domain,
or accept and allow access from requestor to any machine.  The script saves
the user's preferences in a data file which it consults on subsequent
invocations.  If the user has chosen to accept a particular request
automatically, the script returns zero (success) without popping up a dialog
box.
.Sp
Confirmation programs allow the user to manage trust policies when working
with machines that are trusted to different degrees.  For example, a user
might trust the machine on his lan but want to manually confirm requests
from machines in a shared compute cluster.
.IP "sfskey delete \fIkeyname\fR" 4
.IX Item "sfskey delete keyname"
Deletes private key \fIkeyname\fR from the agent (reversing the effect
of an \fBadd\fR command).
.IP "sfskey deleteall" 4
.IX Item "sfskey deleteall"
Deletes all private keys from the agent.
.IP "sfskey edit [\-LP] [\-o \fIkeyname\fR] [\-c \fIcost\fR] [\-l \fIlabel\fR] [\fIkeyname\fR]" 4
.IX Item "sfskey edit [-LP] [-o keyname] [-c cost] [-l label] [keyname]"
Changes the passphrase, passphrase ``cost'', or name of a public key.
Can also download a key from a remote server via \s-1SRP\s0 and store it in a
file.
.Sp
\&\fIkeyname\fR can be a file name, or it can be of the form
\&\fB[\fR\fIuser\fR\fB]@\fR\fIserver\fR, in which case \fBsfskey\fR will
fetch the key remotely and \fIoutfile\fR must be specified.  If
\&\fIkeyname\fR is unspecified the default is \fI$HOME/.sfs/identity\fR.
If \fIkeyname\fR is \fI#\fR, then \fBsfskey edit\fR will search
for the next appropriate keyname in \fI$HOME/.sfs/authkeys\fR.  In this case,
\&\fBsfskey edit\fR will update \fI$HOME/.sfs/identity\fR to point to
this new key by default.
.Sp
The options are:
.RS 4
.IP "\-L" 4
.IX Item "-L"
Does not set symlink in the case that \fIkeyname\fR is \fI#\fR.
.IP "\-P" 4
.IX Item "-P"
Removes any password from the key, so that the password is stored on
disk in unencrypted form.
.IP "\-o \fIkeyname\fR" 4
.IX Item "-o keyname"
Specifies the file to which the edited key should be written.  A
\&\fIkeyname\fR of \fI#\fR implies that \fBsfskey edit\fR should
generate the next available default key in \fI$HOME/.sfs/authkeys\fR.
A \fIkeyname\fR of the form \fIprefix\fR\fI#\fR implies that
\&\fBsfskey edit\fR should generate the next available key in
\&\fI$HOME/.sfs/authkeys\fR with the prefix \fIprefix\fR.  A \fIkeyname\fR
of the form \fIprefix\fR\fI#\fR\fIsuffix\fR implies that 
\&\fBsfskey edit\fR should make a key named 
\&\fI$HOME/.sfs/authkeys/\fR\fIprefix\fR\fI#\fR\fIsuffix\fR.
.IP "\-c \fIcost\fR" 4
.IX Item "-c cost"
Override the default computational cost of processing a password, or
\&\fBPwdCost\fR.
.IP "\-l \fIlabel\fR" 4
.IX Item "-l label"
Specifies the label of the key that shows up in \fBsfskey list\fR.
.RE
.RS 4
.RE
.IP "sfskey 2edit \-[Smp] [\-l \fIlabel\fR] [\-S | \-s \fIsrpfile\fR] [\fIkeyname1\fR \fIkeyname2\fR ...]" 4
.IX Item "sfskey 2edit -[Smp] [-l label] [-S | -s srpfile] [keyname1 keyname2 ...]"
Refreshes a 2\-Schnorr key by re-sharing a secret between a server
and a client. In the case of a compromised client or server, it is 
recommended to refresh a 2\-Schnorr key with this command.  If both
the client and the server have been compromised, a refresh will be 
of little use.
.Sp
Use \fBsfskey 2edit\fR by supplying the keys that you wish to have
updated.  Keynames are given in standard \fBsfskey\fR style.  Keynames
must be either remote keynames (i.e., contain a \fI@\fR but no \fI#\fR
character) or stored in the standard keys directory (i.e., contain a \fI#\fR
but no \fI/\fR character).  For remote keys, \s-1SRP\s0 will be used to 
download the key from the server, and the updated, encrypted client private
keyhalf will be written back to the server along with the new server
keyhalf.  No file will be saved locally.  For keys stored in 
\&\fI$HOME/.sfs/authkeys\fR, \fBsfskey 2edit\fR will update the
server private keyhalf, and write the corresponding client private
keyhalf out to \fI$HOME/.sfs/authkeys\fR under a new filename.  By default,
\&\fBsfskey 2edit\fR will also write the new encrypted client private
keyhalf back to the server for later \s-1SRP\s0 retrieval.
.Sp
If no key is specified, the default key, \fI$HOME/.sfs/identity\fR is 
assumed.
.RS 4
.IP "\-E" 4
.IX Item "-E"
Do not update the encrypted private client key stored on the server.
.IP "\-S" 4
.IX Item "-S"
Do not update \s-1SRP\s0 information on the server.  This option cannot be used
if some of the keynames specified are for remote keys.
.IP "\-m" 4
.IX Item "-m"
Refresh multiple keys.  If you have multiple private splits of the same
private key, this flag will automatically update them all, given that
you've specified one of them.  If you run \fBsfskey 2edit \-m\fR,
with no additional arguments or keynames, \fBsfskey\fR will refresh
all current default keys.
.IP "\-p" 4
.IX Item "-p"
Change password before writing keys out to disk or server.
.IP "\-l \fIlabel\fR" 4
.IX Item "-l label"
Specifies the label of the key that shows up in \fBsfskey list\fR.
.IP "\-s \fIsrpfile\fR" 4
.IX Item "-s srpfile"
Get \s-1SRP\s0 parameters from the file \fIsrpfile\fR.
.RE
.RS 4
.RE
.IP "sfskey gen [\-KP] [\-b \fInbits\fR] [\-c \fIcost\fR] [\-l \fIlabel\fR] [\fIkeyname\fR]" 4
.IX Item "sfskey gen [-KP] [-b nbits] [-c cost] [-l label] [keyname]"
Generates a new Rabin public/private key pair and stores it in \fIkeyname\fR.
It omitted \fIkeyname\fR defaults to the next available Rabin key
in \fI$HOME/.sfs/authkeys\fR.  If \fIkeyname\fR contains a \fI/\fR 
character, it will be treated as a regular Unix file.  If \fIkeyname\fR
is of the form \fIprefix\fR\fI#\fR, \fBsfskey gen\fR will look for
the next available Rabin key in \fI$HOME/.sfs/authkeys\fR with the
prefix \fIprefix\fR.  If \fIkeyname\fR contains a non-terminal \fI#\fR
character, it will be treated as a fully-specified keyname to be saved in 
\&\fI$HOME/.sfs/authkeys\fR.
.Sp
Note that \fBsfskey gen\fR is only useful for generating Rabin keys.
Use either \fBsfskey register\fR or \fBsfskey 2gen\fR to
generate 2\-Schnorr keys.
.RS 4
.IP "\-K" 4
.IX Item "-K"
By default, \fBsfskey gen\fR asks the user to type random text with
which to seed the random number generator.  The \fB\-K\fR option
suppresses that behavior.
.IP "\-P" 4
.IX Item "-P"
Specifies that \fBsfskey gen\fR should not ask for a passphrase and
the new key should be written to disk in unencrypted form.
.IP "\-b \fInbits\fR" 4
.IX Item "-b nbits"
Specifies that the public key should be \fInbits\fR long.
.IP "\-c \fIcost\fR" 4
.IX Item "-c cost"
Override the default computational cost of processing a password, or
\&\fBPwdCost\fR.
.IP "\-l \fIlabel\fR" 4
.IX Item "-l label"
Specifies the label of the key that shows up in \fBsfskey list\fR.
Otherwise, the user will be prompted for a name.
.RE
.RS 4
.RE
.IP "sfskey 2gen [\-BEKP] [\-a {\fIhostid\fR | \-}] [\-b \fInbits\fR] [\-c \fIcost\fR] [\-k \fIokeyname\fR] [\-l \fIlabel\fR] [\-S | \-s \fIsrpfile\fR] [\-w \fIwkeyfile\fR] [\fInkeyname\fR]" 4

agent/sfskey.C
/* $Id: sfskey.C,v 1.96 2004/05/14 23:46:03 max Exp $ */

/*
 *
 * Copyright (C) 1999 David Mazieres (dm@uun.org)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 *
 */

#include "sfskey.h"
#include "srp.h"

bool opt_verbose;
bool opt_quiet;
ref<agentconn> aconn = New refcounted<agentconn> ();

void
nularg (int argc, char **argv)
{
  if (getopt (argc, argv, "") != -1 || optind < argc)
    usage ();
}

void
sfskey_kill (int argc, char **argv)
{
  nularg (argc, argv);
  if (aconn->ccd (false)) {
    int res;
    if (clnt_stat err = aconn->ccd ()->scall (AGENT_KILL, NULL, &res))
      fatal << "sfscd: " << err << "\n";
    else if (res)
      fatal << "sfscd: " << strerror (res) << "\n";
  }
  else if (clnt_stat err = aconn->cagent_ctl ()->scall (AGENTCTL_KILL,
							NULL, NULL))
    if (err != RPC_CANTRECV)
      fatal << "agent: " << err << "\n";
  exit (0);
}

void sfskey_help (int argc, char **argv);
struct modevec {
  const char *name;
  void (*fn) (int argc, char **argv);
  const char *usage;
};
const modevec modes[] = {
  { "add", sfskey_add,
    "add [-t [hrs:]min] [keyfile | [user]@authservhostname]" },
  { "certclear", sfskey_certclear, "certclear" },
  { "certlist", sfskey_certlist, "certlist [-q]" },
  { "certprog", sfskey_certprog,
    "certprog [-p prefixpath] [-f filter] [-e exclude] prog [arg ...]"},
  { "confclear", sfskey_confclear, "confclear" },
  { "conflist", sfskey_conflist, "conflist [-q]" },
  { "confprog", sfskey_confprog, "confprog prog [arg ...]" },
  { "delete", sfskey_delete, "delete keyname" },
  { "del", sfskey_delete, NULL },
  { "deleteall", sfskey_clear, "deleteall" },
  { "delall", sfskey_clear, NULL },
  { "edit", sfskey_edit,
    "edit [-LP] [-o outfile] [-c cost] [-l label] [keyname]" },
  { "gen", sfskey_gen, "gen [-KP] [-b nbits] [-c cost] [-l label] [keyname]" },
  { "generate", sfskey_gen, NULL },
  { "gethash", sfskey_gethash, "gethash [-6] keyname"},
  { "group", sfskey_group, "group [-a key] [-E] [-C] [-L version] "
    "[-m {+|-}membername] [-o {+|-}ownername] groupname" },
  { "help", sfskey_help, "help" },
  { "hostid", sfskey_hostid, "hostid [-s service] {hostname | -}" },
  { "kill", sfskey_kill, "kill" },
  { "list", sfskey_list, "list [-lq]" },
  { "login", sfskey_login, "login [-t [hrs:]min] [user@]hostname" },
  { "logout", sfskey_logout, "logout [user@]{host | realm}" },
  { "ls", sfskey_list, NULL },
  { "norevokeset", sfskey_norevokeset, "norevokeset hostid ... "},
  { "norevokelist", sfskey_norevokelist, "norevokelist"},
  { "passwd", sfskey_passwd, 
    "passwd [pK] [-S | -s srpfile] [-b nbits] [-c cost] [-l label]" },
  { "register", sfskey_reg,
    "register [-fgpPKS] [-b nbits] [-c pwdcost] [-u user] [-l label] "
    "[-w filename] [key]" },
  { "reg", sfskey_reg, NULL },
  { "reset", sfskey_reset, "reset" },
  { "revoke", sfskey_revoke, "revoke {certfile | -}" },
  { "revokegen", sfskey_revokegen, 
    "revokegen [-r newkeyfile [-n newhost]] [-o oldhost] oldkeyfile"},
  { "revokelist", sfskey_revokelist, "revokelist"},
  { "revokeclear", sfskey_revokeclear, "revokeclear"},
  { "revokeprog", sfskey_revokeprog,
    "revokeprog [-b [-f filter] [-e exclude]] prog [arg ...]"},
  { "select", sfskey_select, "select [-f] [key]"},
  { "sesskill", sfskey_sesskill, "sesskill remotehost" },
  { "sesslist", sfskey_sesslist, "sesslist [-q]"},
  { "srpgen", sfskey_srpgen, "srpgen [-b nbits] file" },
  { "srplist", sfskey_srplist, "srplist"},
  { "srpclear", sfskey_srpclear, "srpclear"},
  { "srpcacheprogclear", sfskey_srpcacheprogclear, "srpcacheprogclear" },
  { "srpcacheproglist", sfskey_srpcacheproglist, "srpcacheproglist [-q]" },
  { "srpcacheprog", sfskey_srpcacheprog, "srpcacheprog prog [arg ...]" },
  { "update", sfskey_update,
    "update [-fE] [-S | -s srp_parm_file] [-r srpkey] [-a oldkey] "
    "[-k newkey] server1 [server2 server3 ...]" },
  { "up", sfskey_update, NULL },
  { "user", sfskey_user, "user [ -a key ] username" },
  { "2edit", sfskey_2edit, 
    "2edit [-Emp] [-l label] [-S | -s srpfile] [key1] [key2] ... "},
  { "2gen", sfskey_2gen, 
    "2gen [-BEKP] [-u user] [-b nbits] [-c cost] [-a {server | -}]"
    "[-l label] [-k okeyname] [-S | -s srpfile] [-w wkeyname] [nkeyname]"},
  { NULL, NULL, NULL }
};

static const modevec *sfskey_mode;

void
usage ()
{
  if (!sfskey_mode)
    warnx << "usage: " << progname 
	  << " [-S sock] [-p pwfd] [-vq] command [args]\n"
	  << "	     " << progname << " help\n";
  else {
    while (!sfskey_mode->usage)
      sfskey_mode--;
    warnx << "usage: " << progname << " " << sfskey_mode->usage << "\n";
  }
  exit (1);
}

void
sfskey_help (int argc, char **argv)
{
  strbuf msg;
  msg << "usage: " << progname << " [-S sock] [-p pwfd] command [args]\n";
  for (const modevec *mp = modes; mp->name; mp++)
    if (mp->usage)
      msg << "	 " << progname << " " << mp->usage << "\n";
  make_sync (1);
  msg.tosuio ()->output (1);
  exit (0);
}

static void
sfskey_version ()
{
  strbuf msg;
  msg << "sfskey -- SFS version " << VERSION << "\n";
  make_sync (1);
  msg.tosuio ()->output (1);
  exit (0);
}


int
main (int argc, char **argv)
{
  setprogname (argv[0]);
  xputenv ("POSIXLY_CORRECT=1"); // Prevents Linux from reordering options
  sfsconst_init ();
  srp_base::minprimsize = sfs_mindlogsize;

  int ch;
  while ((ch = getopt (argc, argv, "S:p:vqV")) != -1)
    switch (ch) {
    case 'S':
      agentsock = optarg;
      break;
    case 'V':
      sfskey_version ();
      break;
    case 'p':
      {
	int fd;
	if (!convertint (optarg, &fd))
	  usage ();
	opt_pwd_fd = true;
	close_on_exec (fd);	// Paranoia
	pwd_fds.push_back (fd);
	break;
      }
    case 'v':
      opt_verbose = true;
      break;
    case 'q':
      opt_quiet = true;
      break;
    default:
      usage ();
      break;
    }
  if (optind >= argc)
    usage ();

  const modevec *mp;
  for (mp = modes; mp->name; mp++)
    if (!strcmp (argv[optind], mp->name))
      break;
  if (!mp->name)
    usage ();
  sfskey_mode = mp;

  optind++;

  mp->fn (argc, argv);
  amain ();
}

#ifdef XXX_EXIT
#undef exit
/* XXX - gcc-2.95.2 bug on alpha */
void
XXX_call_exit (int code)
{
  exit (code);
}
#endif /* XXX_EXIT */
